File: src/test/scala/com/exasol/cloudetl/scriptclasses/FilesDataImporterTest.scala
package com.exasol.cloudetl.scriptclasses

import java.nio.file.Files
import java.nio.file.Paths

import scala.io.Source

import com.exasol.ExaIterator
import com.exasol.ExaMetadata
import com.exasol.cloudetl.storage.StorageProperties
import com.exasol.parquetio.data.ChunkInterval
import com.exasol.parquetio.data.ChunkIntervalImpl

import org.mockito.Mockito._
import org.scalatest.matchers.should._

class FilesDataImporterTest extends StorageTest with Matchers {

  private[this] val properties = Map(
    "BUCKET_PATH" -> testResourceParquetPath,
    "DATA_FORMAT" -> "PARQUET"
  )

  test("run throws if file format is not supported") {
    val file = s"$testResourceDir/import/parquet/sales_positions1.snappy.parquet"
    val iter = mockFileIterator("CSV", file)
    val thrown = intercept[IllegalArgumentException] {
      FilesDataImporter.run(mock[ExaMetadata], iter)
    }
    assert(thrown.getMessage().startsWith("E-CSE-17"))
    assert(thrown.getMessage().contains("file format 'CSV' is not supported."))
  }

  test("run emits all records from a source") {
    val file1 = s"$testResourceDir/import/parquet/sales_positions1.snappy.parquet"
    val file2 = s"$testResourceDir/import/parquet/sales_positions2.snappy.parquet"
    val expectedNumberOfRecords = 1000

    val storageProperties = StorageProperties(properties)
    val iter =
      exaIteratorMock(iteratorRow(storageProperties, file1, 0, 1), iteratorRow(storageProperties, file2, 0, 1))

    FilesDataImporter.run(mock[ExaMetadata], iter)
    assert(iter.getEmittedRows().size() == expectedNumberOfRecords)
  }

  test("run emits records for single interval") {
    val file1 = s"$testResourceDir/import/parquet/sales_positions1.snappy.parquet"
    val storageProperties = StorageProperties(properties)
    val iter = exaIteratorMock(iteratorRow(storageProperties, file1, 0, 1))
    FilesDataImporter.run(mock[ExaMetadata], iter)
    assert(iter.getEmittedRows().size() == 500)
  }

  test("run emits records for duplicate intervals") {
    val file = s"$testResourceDir/import/parquet/sales_positions1.snappy.parquet"
    val storageProperties = StorageProperties(properties)
    val iter =
      exaIteratorMock(iteratorRow(storageProperties, file, 0, 1), iteratorRow(storageProperties, file, 0, 1))
    FilesDataImporter.run(mock[ExaMetadata], iter)
    assert(iter.getEmittedRows().size() == 500)
  }

  /**
   * +---------+-----------+----------+------+-----+----------+--------+
   * |sales_id |position_id|article_id|amount|price|voucher_id|canceled|
   * +---------+-----------+----------+------+-----+----------+--------+
   * |582244536|2          |96982     |1     |0.56 |null      |null    |
   * |582177839|6          |96982     |2     |0.56 |null      |null    |
   * |582370207|0          |96982     |1     |0.56 |null      |null    |
   * |582344312|0          |96982     |5     |0.56 |null      |null    |
   * |582344274|1          |96982     |1     |0.56 |null      |null    |
   * +---------+-----------+----------+------+-----+----------+--------+
   */
  test("run emits correct sequence of records from PARQUET file") {
    val parquetFile = s"$testResourceDir/import/parquet/sales_positions_small.snappy.parquet"
    val iter = mockFileIterator("PARQUET", parquetFile)
    FilesDataImporter.run(mock[ExaMetadata], iter)
    verifySmallFilesImport(iter)
  }

  test("run emits correct sequence of records from AVRO file") {
    val avroFile = s"$testResourceDir/import/avro/sales_positions_small.avro"
    val iter = mockFileIterator("AVRO", avroFile)
    FilesDataImporter.run(mock[ExaMetadata], iter)
    verifySmallFilesImport(iter)
  }

  test("collectFiles for empty iterator (will never happen in a UDF)") {
    val result = FilesDataImporter.collectFiles(ExaIteratorMock.empty())
    assert(result.size == 1)
    result.get(null).get.should(contain).theSameElementsAs(List(chunk(0, 0)))
  }

  test("collectFiles for iterator with single entry") {
    val result =
      FilesDataImporter.collectFiles(exaIteratorMock(iteratorRow("file1.parquet", 17L, 42L)))
    result.get("file1.parquet").get.should(contain).theSameElementsAs(List(chunk(17, 42)))
  }

  test("collectFiles for iterator with single file but multiple chunks") {
    val result =
      FilesDataImporter.collectFiles(
        exaIteratorMock(
          iteratorRow("file1.parquet", 17L, 42L),
          iteratorRow("file1.parquet", 1L, 2L)
        )
      )
    assert(result.size == 1)
    result.get("file1.parquet").get.should(contain).theSameElementsAs(List(chunk(17, 42), chunk(1, 2)))
  }

  test("collectFiles for iterator with multiple files and multiple chunks") {
    val result =
      FilesDataImporter.collectFiles(
        exaIteratorMock(
          iteratorRow("file1.parquet", 17L, 42L),
          iteratorRow("file1.parquet", 1L, 2L),
          iteratorRow("file2.parquet", 0L, 1L)
        )
      )
    assert(result.size == 2)
    result.get("file1.parquet").get.should(contain).theSameElementsAs(List(chunk(17, 42), chunk(1, 2)))
    result.get("file2.parquet").get.should(contain).theSameElementsAs(List(chunk(0, 1)))
  }

  test("collectFiles for iterator with two files") {
    val result =
      FilesDataImporter.collectFiles(
        exaIteratorMock(
          iteratorRow("file1.parquet", 17L, 42L),
          iteratorRow("file2.parquet", 1L, 2L)
        )
      )
    assert(result.size == 2)
    result.get("file1.parquet").get.should(contain).theSameElementsAs(List(chunk(17, 42)))
    result.get("file2.parquet").get.should(contain).theSameElementsAs(List(chunk(1, 2)))
  }

  test("collectFiles with 40k files") {
    val files = (0 until 40000).map(i => iteratorRow(s"file$i.parquet", 0L, 1L))
    val result = FilesDataImporter.collectFiles(ExaIteratorMock.fromSeq(files))
    assert(result.size == 40000)
    result.values.foreach(intervals => assert(intervals.size() == 1))
  }

  private def exaIteratorMock(files: Array[Object]*) = ExaIteratorMock.fromSeq(files)

  private def iteratorRow(file: String, intervalStart: java.lang.Long, intervalEnd: java.lang.Long) =
    Array[Object](null, null, file, intervalStart, intervalEnd)

  private def iteratorRow(
    storageProperties: StorageProperties,
    file: String,
    intervalStart: java.lang.Long,
    intervalEnd: java.lang.Long
  ) =
    Array[Object](storageProperties.getStoragePath(), storageProperties.mkString(), file, intervalStart, intervalEnd)

  private[this] def chunk(start: Long, end: Long): ChunkInterval =
    new ChunkIntervalImpl(start, end)

  private[this] def mockFileIterator(fileFormat: String, filename: String): ExaIterator = {
    val iter = mockExasolIterator(properties ++ Map("DATA_FORMAT" -> fileFormat))
    when(iter.next()).thenReturn(false)
    when(iter.getString(2)).thenReturn(filename)
    when(iter.getLong(3)).thenReturn(0L)
    when(iter.getLong(4)).thenReturn(1L)
    iter
  }

  private[this] def verifySmallFilesImport(iter: ExaIterator): Unit = {
    val totalRecords = 5
    val records: Seq[Seq[Object]] = Seq(
      Seq[Any](582244536L, 2, 96982, 1, 0.56, null, null),
      Seq[Any](582177839L, 6, 96982, 2, 0.56, null, null),
      Seq[Any](582370207L, 0, 96982, 1, 0.56, null, null),
      Seq[Any](582344312L, 0, 96982, 5, 0.56, null, null),
      Seq[Any](582344274L, 1, 96982, 1, 0.56, null, null)
    ).map { seq =>
      seq.map(_.asInstanceOf[AnyRef])
    }

    verify(iter, times(totalRecords)).emit(anyObjects())
    records.foreach { case rows =>
      verify(iter, times(1)).emit(rows: _*)
    }
  }

}

File: src/test/java/com/exasol/cloudetl/scriptclasses/FilesDataImporterTest.java
package com.exasol.cloudetl.scriptclasses;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import org.junit.jupiter.api.Test;

import com.exasol.*;
import com.exasol.cloudetl.ScalaConverters;
import com.exasol.cloudetl.storage.StorageProperties;
import com.exasol.parquetio.data.*;

class FilesDataImporterTest extends StorageTest {
    private final Map<String, String> properties = Map.of("BUCKET_PATH", this.testResourceParquetPath, "DATA_FORMAT",
            "PARQUET");

    @Test
    void runThrowsIfFileFormatIsNotSupported() throws Exception {
        final String file = this.testResourceDir + "/import/parquet/sales_positions1.snappy.parquet";
        final ExaIterator iter = mockFileIterator("CSV", file);
        final IllegalArgumentException thrown = assertThrows(IllegalArgumentException.class,
                () -> FilesDataImporter.run(mock(ExaMetadata.class), iter));
        assertTrue(thrown.getMessage().startsWith("E-CSE-17"));
        assertTrue(thrown.getMessage().contains("file format 'CSV' is not supported."));
    }

    @Test
    void runEmitsAllRecordsFromASource() throws Exception {
        final String file1 = this.testResourceDir + "/import/parquet/sales_positions1.snappy.parquet";
        final String file2 = this.testResourceDir + "/import/parquet/sales_positions2.snappy.parquet";
        final StorageProperties storageProperties = new StorageProperties(this.properties);
        final ExaIteratorMock iter = exaIteratorMock(iteratorRow(storageProperties, file1, 0L, 1L),
                iteratorRow(storageProperties, file2, 0L, 1L));
        FilesDataImporter.run(mock(ExaMetadata.class), iter);
        assertEquals(1000, iter.getEmittedRows().size());
    }

    @Test
    void runEmitsRecordsForSingleInterval() throws Exception {
        final String file1 = this.testResourceDir + "/import/parquet/sales_positions1.snappy.parquet";
        final StorageProperties storageProperties = new StorageProperties(this.properties);
        final ExaIteratorMock iter = exaIteratorMock(iteratorRow(storageProperties, file1, 0L, 1L));
        FilesDataImporter.run(mock(ExaMetadata.class), iter);
        assertEquals(500, iter.getEmittedRows().size());
    }

    @Test
    void runEmitsRecordsForDuplicateIntervals() throws Exception {
        final String file = this.testResourceDir + "/import/parquet/sales_positions1.snappy.parquet";
        final StorageProperties storageProperties = new StorageProperties(this.properties);
        final ExaIteratorMock iter = exaIteratorMock(iteratorRow(storageProperties, file, 0L, 1L),
                iteratorRow(storageProperties, file, 0L, 1L));
        FilesDataImporter.run(mock(ExaMetadata.class), iter);
        assertEquals(500, iter.getEmittedRows().size());
    }

    @Test
    void runEmitsCorrectSequenceOfRecordsFromParquetFile() throws Exception {
        final String parquetFile = this.testResourceDir + "/import/parquet/sales_positions_small.snappy.parquet";
        final ExaIterator iter = mockFileIterator("PARQUET", parquetFile);
        FilesDataImporter.run(mock(ExaMetadata.class), iter);
        verifySmallFilesImport(iter);
    }

    @Test
    void runEmitsCorrectSequenceOfRecordsFromAvroFile() throws Exception {
        final String avroFile = this.testResourceDir + "/import/avro/sales_positions_small.avro";
        final ExaIterator iter = mockFileIterator("AVRO", avroFile);
        FilesDataImporter.run(mock(ExaMetadata.class), iter);
        verifySmallFilesImport(iter);
    }

    @Test
    void collectFilesForEmptyIterator() throws Exception {
        final Map<String, java.util.List<ChunkInterval>> result = collectFiles(ExaIteratorMock.empty());
        assertEquals(1, result.size());
        assertChunks(result.get(null), chunk(0, 0));
    }

    @Test
    void collectFilesForIteratorWithSingleEntry() throws Exception {
        final Map<String, java.util.List<ChunkInterval>> result = collectFiles(exaIteratorMock(iteratorRow("file1.parquet", 17L, 42L)));
        assertChunks(result.get("file1.parquet"), chunk(17, 42));
    }

    @Test
    void collectFilesForIteratorWithSingleFileButMultipleChunks() throws Exception {
        final Map<String, java.util.List<ChunkInterval>> result = collectFiles(
                exaIteratorMock(iteratorRow("file1.parquet", 17L, 42L), iteratorRow("file1.parquet", 1L, 2L)));
        assertEquals(1, result.size());
        assertChunks(result.get("file1.parquet"), chunk(17, 42), chunk(1, 2));
    }

    @Test
    void collectFilesForIteratorWithMultipleFilesAndMultipleChunks() throws Exception {
        final Map<String, java.util.List<ChunkInterval>> result = collectFiles(exaIteratorMock(
                iteratorRow("file1.parquet", 17L, 42L), iteratorRow("file1.parquet", 1L, 2L),
                iteratorRow("file2.parquet", 0L, 1L)));
        assertEquals(2, result.size());
        assertChunks(result.get("file1.parquet"), chunk(17, 42), chunk(1, 2));
        assertChunks(result.get("file2.parquet"), chunk(0, 1));
    }

    @Test
    void collectFilesForIteratorWithTwoFiles() throws Exception {
        final Map<String, java.util.List<ChunkInterval>> result = collectFiles(
                exaIteratorMock(iteratorRow("file1.parquet", 17L, 42L), iteratorRow("file2.parquet", 1L, 2L)));
        assertEquals(2, result.size());
        assertChunks(result.get("file1.parquet"), chunk(17, 42));
        assertChunks(result.get("file2.parquet"), chunk(1, 2));
    }

    @Test
    void collectFilesWith40kFiles() throws Exception {
        final List<Object[]> files = IntStream.range(0, 40_000)
                .mapToObj(index -> iteratorRow("file" + index + ".parquet", 0L, 1L)).collect(Collectors.toList());
        final Map<String, java.util.List<ChunkInterval>> result = collectFiles(ExaIteratorMock.fromList(files));
        assertEquals(40_000, result.size());
        result.values().forEach(intervals -> assertEquals(1, intervals.size()));
    }

    private ExaIteratorMock exaIteratorMock(final Object[]... files) {
        return ExaIteratorMock.fromList(Arrays.asList(files));
    }

    private Object[] iteratorRow(final String file, final Long intervalStart, final Long intervalEnd) {
        return new Object[] { null, null, file, intervalStart, intervalEnd };
    }

    private Object[] iteratorRow(final StorageProperties storageProperties, final String file, final Long intervalStart,
            final Long intervalEnd) {
        return new Object[] { storageProperties.getStoragePath(), storageProperties.mkString(), file, intervalStart,
                intervalEnd };
    }

    private ChunkInterval chunk(final long start, final long end) {
        return new ChunkIntervalImpl(start, end);
    }

    private ExaIterator mockFileIterator(final String fileFormat, final String filename) throws Exception {
        final Map<String, String> params = new HashMap<>(this.properties);
        params.put("DATA_FORMAT", fileFormat);
        final ExaIterator iter = mockExasolIterator(params);
        when(iter.next()).thenReturn(false);
        when(iter.getString(2)).thenReturn(filename);
        when(iter.getLong(3)).thenReturn(0L);
        when(iter.getLong(4)).thenReturn(1L);
        return iter;
    }

    private Map<String, java.util.List<ChunkInterval>> collectFiles(final ExaIterator iter) throws Exception {
        return ScalaConverters.javaMapCopy(FilesDataImporter.collectFiles(iter));
    }

    private void assertChunks(final java.util.List<ChunkInterval> actual, final ChunkInterval... expected) {
        assertNotNull(actual);
        assertEquals(expected.length, actual.size());
        assertTrue(actual.containsAll(Arrays.asList(expected)));
    }

    private void verifySmallFilesImport(final ExaIterator iter) throws Exception {
        final List<List<Object>> records = List.of(//
                Arrays.asList(582244536L, 2, 96982, 1, 0.56, null, null), //
                Arrays.asList(582177839L, 6, 96982, 2, 0.56, null, null), //
                Arrays.asList(582370207L, 0, 96982, 1, 0.56, null, null), //
                Arrays.asList(582344312L, 0, 96982, 5, 0.56, null, null), //
                Arrays.asList(582344274L, 1, 96982, 1, 0.56, null, null));

        verify(iter, times(5)).emit(anyObjects());
        for (final List<Object> row : records) {
            verify(iter, times(1)).emit(row.toArray());
        }
    }
}
